src/scipp/compat/xarray_compat.py

```diff
@@ -30,71 +29,128 @@ def from_xarray(obj: Union[xr.DataArray, xr.Dataset]) -> VariableLike:
+        return _to_xarray_variable(obj)
+    elif isinstance(obj, DataArray):
+        return _to_xarray_dataarray(obj)
+    elif isinstance(obj, Dataset):
+        return _to_xarray_dataset(obj)
+    else:
+        raise ValueError(f"to_xarray: cannot convert type '{type(obj)}'")
+
+
+def _from_xarray_variable(xr_obj: Union[xr.Coordinate, xr.DataArray]) -> Variable:
+    """Converts an xarray Coordinate or the data in a DataArray to a scipp.Variable.
+    """
+    unit = xr_obj.attrs.get('units', None)
+    return Variable(dims=xr_obj.dims,
+                    values=xr_obj.values,
+                    unit=Unit(unit) if unit is not None else default_unit)
+
+
+def _to_xarray_variable(var: Variable) -> xr.Variable:
+    """Converts a scipp.Variable to a dict containing dims, values and unit for storing
+    in either an xarray Coordinate or DataArray.
+    """
+    import xarray as xr
+    if var.bins is not None:
+        raise ValueError("Xarray does not support binned data.")
+    if var.variances is not None:
+        warn("Variances of variable were stripped when converting to Xarray.")
+    attrs = {'units': str(var.unit)} if var.unit is not None else None
+    return xr.Variable(dims=var.dims, data=var.values, attrs=attrs)
+
+
+def _from_xarray_dataarray(da: xr.DataArray) -> DataArray:
+    """Converts an xarray.DataArray object to a scipp.DataArray object.
     """
     coords = {}
     attrs = {attr: scalar(da.attrs[attr]) for attr in da.attrs if attr != "units"}
 
     for name, coord in da.coords.items():
         if name in da.indexes:
-            coords[name] = _var_from_xarray(coord)
+            coords[name] = _from_xarray_variable(coord)
         else:
-            attrs[f"{name}"] = _var_from_xarray(coord)
+            attrs[f"{name}"] = _from_xarray_variable(coord)
 
-    return DataArray(data=_var_from_xarray(da),
+    return DataArray(data=_from_xarray_variable(da),
                      coords=coords,
                      attrs=attrs,
                      name=da.name or "")
 
 
-def from_xarray_dataset(ds: xr.Dataset) -> Dataset:
-    """Converts an xarray.Dataset object to a scipp.Dataset object.
+def _to_xarray_dataarray(da: DataArray) -> xr.DataArray:
+    """Converts a scipp.DataArray object to an xarray.DataArray object.
+    """
+    import xarray as xr
+    if da.masks:
+        warn("Some masks were found in the DataArray. "
+             "These have been removed when converting to Xarray.")
+    data = _to_xarray_variable(da.data)
+    coords = {}
+    for key, coord in {**da.coords, **da.attrs}.items():
+        for dim in coord.dims:
+            if da.meta.is_edges(key, dim=dim):
+                raise ValueError("Xarray does not support coordinates with bin edges.")
+        coords[key] = _to_xarray_variable(coord)
 
-    Parameters
-    ----------
-    ds:
-        An xarray.Dataset object to be converted.
+    return xr.DataArray(data, coords={key: coord for key, coord in coords.items()})
```

Comment:
```suggestion
    return xr.DataArray(data, coords=coords)
```
?

Also, why don't you just add the coords directly, instead of making a temporary dict first?